### PR TITLE
[CUDA] Allow async copy for fp16

### PR DESCRIPTION
**Pull request by nirvedhmeshram**
Async copy can be used to form better software pipeline and reduce the pressure on register use in fp16 too.

### CODE REVIEW

**Review by MaheshRavishankar — APPROVED** [review]
Looks OK to me. Dont know if you have any numbers to show how well these works/ wait for Thomas to take a look as well.

**Comment by nirvedhmeshram** [comment]
For numbers, I just see register use and kernel time dropping in ncu, and from correctness point of view did not notice a change in any gemms. 
before
image
after
image

**Comment by nirvedhmeshram** [comment]
I added copy of 8 elements for fp16 instead of 4 that is done for fp32, interestingly the software pipline doesnt quite seem to work after this change, even when I give a pipeline depth of 4, it seems to give output equivalent to pipeline of 1, I will try to debug tomorrow

**Comment by nirvedhmeshram** [comment]
There was no real issue, I just had to double the default k tile size to account for copying chunks of 8 rather than 4 to generate code that is similar to fp32 code

File: compiler/src/iree/compiler/Codegen/Common/GPUDistributeSharedMemoryCopy.cpp [hunk 1]
```diff
@@ -39,15 +39,21 @@ static void populateTilingCopyToWorkgroupMemPatterns(
...
+        int64_t copyTileSize;
         // We tile to 4 as we want each thread to load 4 element in a cyclic
         // distribution.
         SmallVector<Value, 4> tileSizesVal;
-        unsigned rank = cast<linalg::GenericOp>(operation)
-                            .getOperand(0)
-                            .getType()
-                            .cast<MemRefType>()
-                            .getRank();
+        MemRefType lhsMemRefType = cast<linalg::GenericOp>(operation)
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/GPUDistributeSharedMemoryCopy.cpp:46]: Is it safe to say you are trying a 128 bit copy. So maybe just do 128 / (element type bitwidth) here?

File: compiler/src/iree/compiler/Codegen/Common/GPUDistributeSharedMemoryCopy.cpp [hunk 2]
```diff
@@ -101,6 +104,12 @@ static Optional<SmallVector<int64_t, 4>> getGPUNativeVectorSize(
   if (!vt) return llvm::None;
   if (!vt.permutation_map().isMinorIdentity()) return llvm::None;
   ArrayRef<int64_t> shape = vt.getVectorType().getShape();
+  int targetVectorSize;
+  if (vt.getVectorType().getElementType().isF16()) {
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/GPUDistributeSharedMemoryCopy.cpp:108]: Same goes here. vector size seems to be 128 / bitwidth

File: compiler/src/iree/compiler/Codegen/Common/GPUDistributeSharedMemoryCopy.cpp [hunk 3]
```diff
@@ -158,6 +167,12 @@ static void distributeTransferRead(func::FuncOp funcOp, Value flatThreadId,
     Value id = flatThreadId;
     SmallVector<int64_t, 2> multiplier;
     auto shape = readOp.getVectorType().getShape();
+    int targetVectorSize = 4;
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/GPUDistributeSharedMemoryCopy.cpp:170]: Maybe this needs to be moved into a utility function.

**Review by MaheshRavishankar — COMMENTED** [review]
Ill let Thomas review more. Left a few comments.

File: compiler/src/iree/compiler/Codegen/Common/GPUDistributeSharedMemoryCopy.cpp (hunk 1 above)
Comment (nirvedhmeshram) [reply compiler/src/iree/compiler/Codegen/Common/GPUDistributeSharedMemoryCopy.cpp:46]: Makes sense, I am just unsure, how to accomplish the 128 bit copy with the existing functions

File: compiler/src/iree/compiler/Codegen/Common/GPUDistributeSharedMemoryCopy.cpp (hunk 3 above)
Comment (nirvedhmeshram) [reply compiler/src/iree/compiler/Codegen/Common/GPUDistributeSharedMemoryCopy.cpp:170]: Makes sense, @ThomasRaoux  also suggested making this choice of 4 for fp32 and 8 for fp16 in one place only, I will discuss how best to do that with him tomorrow.

File: compiler/src/iree/compiler/Codegen/Common/GPUDistributeSharedMemoryCopy.cpp (hunk 1 above)
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/Common/GPUDistributeSharedMemoryCopy.cpp:46]: +1 to what Mahesh said, you can use `getElementTypeBitWidth` to get the bit size and deduce the number of elements based on that. And we can have a common function to get the value across the different cases.

**Review by ThomasRaoux — APPROVED** [review]

**Pull request merged**